### PR TITLE
Return empty hash if nil address

### PR DESCRIPTION
**Pull request by data-doge**
## Description of change

While testing the COE form, we noticed that calls to `GET https://staging-api.va.gov/v0/in_progress_forms/26-1880` were 500'ing. This is because `User#address` was throwing errors when there was no address on `identity` and `mpi_profile`. See https://github.com/department-of-veterans-affairs/vets-api/blob/0ec4937c741bf2c3e3e6af32116234fb744dc5d2/app/models/user.rb#L180. Consequently, the COE form could not be autofilled with information that the user previously specified. It's possible that other forms are experiencing this issue as well.

## Original issue(s)

https://github.com/department-of-veterans-affairs/va.gov-team/issues/47435. This PR is related to this issue only because the bug was discovered while testing resolutions to the issues presented there. I don't believe this issue is COE-specific, though, it's necessary to resolve the issue for COE to work as intended.

## Things to know about this PR

See above.

### CODE REVIEW

**Comment by kylesoskin** [comment]
Hey @data-doge, I will look at this tomorrow. 

There is probably a million different things using the user model so might want to poke around a bit and make sure nothing any layers up in the code is relying on that failing if empty for any reason (although doing that is not great, and should be changed any way in my opinion, or done a different way, still probably best just to check/confirm. 


Like make sure there isn't anything like 

begin
  Pdf.stamp(user.address)
rescue
  raise "Error: address not found, cannot generate form"
end

Not saying there is that, but should double check. I have seen some stuff like that in the code in places and there is a lot of relying on begin/rescues here in this backend code for some reason.... 

Other than that it is probably fine. Although there is another method Ive seen used that does what you want already that might be better than the additional "or"

**Comment by data-doge** [comment]
Nice, thanks @kylesoskin! I'll do some more digging around.

File: app/models/user.rb [hunk 1]
```diff
@@ -175,7 +175,7 @@ def active_mhv_ids
   end
 
   def address
-    address = identity&.address || mpi_profile&.address
+    address = identity&.address || mpi_profile&.address || {}
```
Comment (bosawt) [inline app/models/user.rb:178]: Can you wrap this in an `OpenStruct` and then refer to the fields as functions? As in:

`address = OpenStruct.new(identity&.address || mpi_profile&.address)`
`{ street: address.street, street2: address.street2 ...`

I think that more closely will represent what we intended here, as in, to always return a hash with fields, even if they are empty (this is how the `identity.address` or `mpi_profile.address` functions return their address objects). It should also solve the case when the identity address object or MPI address object was somehow never defined in the first place, which is what you're trying to accomplish here

Comment (bosawt) [reply app/models/user.rb:178]: Actually I suppose an empty hash accomplishes the same thing, for some reason I was thinking it was returning `nil`, instead this is functionally doing exactly what I asked above, haha.

In that case either implementation should be safe, it was unintentional that calling `user.address` would return an error, even if address wasn't defined at all

Comment (kylesoskin) [reply app/models/user.rb:178]: Yea that is what I was referring to in my comment 
>Although there is another method Ive seen used that does what you want already that might be better than the additional "or"

I hadn't had time to dig that up. Thanks!!

Comment (kylesoskin) [reply app/models/user.rb:178]: Actually, yea, you are right, this is fine I think as is.

**Review by kylesoskin — DISMISSED** [review]
Approved/looks good to me, assuming nothing else relies on this code erroring when there is no address

**Review by bosawt — DISMISSED** [review]

**Review by holdenhinkle — APPROVED** [review]
Do you want to do any logging when `user.address` throws an error?

Otherwise, LGTM.

**Comment by data-doge** [comment]
Hey @bosawt @kylesoskin, thanks for the review! After Kyle's comment, I surveyed the codebase to check whether anything depended on `User#address` to raise an error. I found a couple files where that _might_ be the case. I've listed them in the commit message here: https://github.com/department-of-veterans-affairs/vets-api/pull/11145/commits/72b186becbe2c623f7c07bd6f10bfbde1f6a0965. In that commit, I've moved this code from the `User` model into the `FormProfile` to make sure that I'm not screwing any other code up. 

@bosawt, it's helpful to know that `User#address` was not intended to raise errors. Do you think we can merge this updated PR, and then do you think that the VSP Identity team can look into the files mentioned above and potentially re-implement my prior change at their discretion? For context, the Benefits Team is focused on fixing bugs quickly to unblock the launch of a new form. Thanks!

**Comment by data-doge** [comment]
Thanks for the review @holdenhinkle! AFAICT, and especially in light of @bosawt's [recent comment](https://github.com/department-of-veterans-affairs/vets-api/pull/11145#discussion_r1023089928), it seems that it is perfectly normal for a user to not have an address. Given that, I think we can refrain from logging here.

**Comment by kylesoskin** [comment]
In that latest commit it will now return `{}` whereas previously it would have returned 
```
{
      street: nil,
      street2: nil,
      city: nil,
      state: nil,
      country: nil,
      postal_code: nil
}
```

 Not sure if that matters or if either is any better than the other.... but, just saying.

**Comment by bosawt** [comment]
> Hey @bosawt @kylesoskin, thanks for the review! After Kyle's comment, I surveyed the codebase to check whether anything depended on `User#address` to raise an error. I found a couple files where that _might_ be the case. I've listed them in the commit message here: [72b186b](https://github.com/department-of-veterans-affairs/vets-api/commit/72b186becbe2c623f7c07bd6f10bfbde1f6a0965). In that commit, I've moved this code from the `User` model into the `FormProfile` to make sure that I'm not screwing any other code up.
> 
> @bosawt, it's helpful to know that `User#address` was not intended to raise errors. [Ref](https://github.com/department-of-veterans-affairs/vets-api/pull/11145#discussion_r1023089928). Do you think we can merge this updated PR, and then do you think that the VSP Identity team can look into the files mentioned above and potentially re-implement my prior change at their discretion? For context, the Benefits Team is focused on fixing bugs quickly to unblock the launch of a new form. Thanks!

Thanks for being thorough and looking through possible issues from changing the User model, I did look through those just to confirm, but just wanted you to know that the previous change is fine too. Of course, I won't stop you with the current PR, but if you preferred the solution by changing the User model I'll approve it.

Generally I think it's unexpected when attributes on the User model raise errors, so I prefer the User model to return nil, and then the services to contextually decide whether or not that's worthy of stopping the process for whatever service they're trying to provide

**Comment by data-doge** [comment]
Thanks again both. And sounds good @bosawt! In that case, I'll gladly go back to the original commit.

**Review by bosawt — APPROVED** [review]
Looks good, thanks for adding specs too!

**Pull request merged**